### ISSUE TITLE
REST Api order line_items variable type is not the same as the expected variable type

### ISSUE THREAD

**Issue by alexbouchardd** [woocommerce/woocommerce#17766]
It appears that 3.2.2 or 3.2.4 has introduce a bug where the response format for a order line items does not match the expected format.

I am having issues with the following fields:
```
'tax_class' => array(
  'description' => __( 'Tax class of product.', 'woocommerce' ),
  'type'        => 'integer',
  'context'     => array( 'view', 'edit' ),
),
'price' => array(
  'description' => __( 'Product price.', 'woocommerce' ),
  'type'        => 'string',
  'context'     => array( 'view', 'edit' ),
  'readonly'    => true,
),
```
The line_items returned by the api have `tax_class` as a `string` and `price` as a `integer`.

The problem arise when updating an order. By adding a new line item and updating the line items the previously returned line_items causing a 400 `code: 'rest_invalid_param'`.

For the sake of experimentation I changed the `type` of `price` and `tax_class` to `integer` and `string` respectively and manage to make it work without issues. However, I am unaware if such a change had side effects.

First time reporting an issue here, feel free to let me know if there is anything I need to add!

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Get a specific order via the rest api GET /wp-json/wc/v2/orders/<id>
2. Add a line item to the order
3. Update the order POST /wp-json/wc/v2/orders/<id>

## Expected/actual behavior

An order should return the same variable type then the rest api expects to receive.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>

### WordPress Environment ###
Home URL: https://lecoffretderachel.com
Site URL: https://fromrachel.com
WC Version: 3.2.4
Log Directory Writable: ✔
WP Version: 4.9
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: –
WP Cron: ✔
Language: fr_FR

### Server Environment ###

Server Info: Apache/2.4.28 (Unix) OpenSSL/1.0.2g
PHP Version: 7.0.25
PHP Post Max Size: 1 GB
PHP Time Limit: 3600
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.20
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.2.4
WC Database Prefix: wp_37gfanz9jp_
MaxMind GeoIP Database: ✔
Total Database Size: 797.01MB
Database Data Size: 559.05MB
Database Index Size: 237.96MB
wp_37gfanz9jp_woocommerce_sessions: Data: 127.02MB + Index: 3.02MB
wp_37gfanz9jp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_37gfanz9jp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_37gfanz9jp_woocommerce_order_items: Data: 7.52MB + Index: 3.52MB
wp_37gfanz9jp_woocommerce_order_itemmeta: Data: 44.58MB + Index: 36.11MB
wp_37gfanz9jp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_37gfanz9jp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_37gfanz9jp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
wp_37gfanz9jp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_woocommerce_payment_tokens: Data: 0.30MB + Index: 0.14MB
wp_37gfanz9jp_woocommerce_payment_tokenmeta: Data: 1.52MB + Index: 0.84MB
wp_37gfanz9jp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_commentmeta: Data: 11.11MB + Index: 1.59MB
wp_37gfanz9jp_comments: Data: 51.58MB + Index: 33.98MB
wp_37gfanz9jp_gf_addon_feed: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_cms_nav_cache: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_icl_content_status: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_core_status: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_flags: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_languages: Data: 0.02MB + Index: 0.03MB
wp_37gfanz9jp_icl_languages_translations: Data: 0.19MB + Index: 0.08MB
wp_37gfanz9jp_icl_locale_map: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_message_status: Data: 0.02MB + Index: 0.03MB
wp_37gfanz9jp_icl_mo_files_domains: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_node: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_icl_reminders: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_icl_strings: Data: 6.52MB + Index: 8.55MB
wp_37gfanz9jp_icl_string_packages: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_icl_string_pages: Data: 3.52MB + Index: 2.52MB
wp_37gfanz9jp_icl_string_positions: Data: 1.52MB + Index: 0.14MB
wp_37gfanz9jp_icl_string_status: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_string_translations: Data: 5.52MB + Index: 2.52MB
wp_37gfanz9jp_icl_string_urls: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_translate: Data: 3.52MB + Index: 0.11MB
wp_37gfanz9jp_icl_translate_job: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_icl_translations: Data: 1.52MB + Index: 3.70MB
wp_37gfanz9jp_icl_translation_batches: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_icl_translation_status: Data: 3.52MB + Index: 0.08MB
wp_37gfanz9jp_links: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_options: Data: 99.19MB + Index: 2.09MB
wp_37gfanz9jp_postmeta: Data: 116.64MB + Index: 92.25MB
wp_37gfanz9jp_posts: Data: 19.53MB + Index: 7.88MB
wp_37gfanz9jp_rg_form: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_rg_form_meta: Data: 0.11MB + Index: 0.00MB
wp_37gfanz9jp_rg_form_view: Data: 0.08MB + Index: 0.09MB
wp_37gfanz9jp_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_rg_lead: Data: 2.14MB + Index: 0.09MB
wp_37gfanz9jp_rg_lead_detail: Data: 9.06MB + Index: 10.14MB
wp_37gfanz9jp_rg_lead_detail_long: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_rg_lead_meta: Data: 1.05MB + Index: 1.17MB
wp_37gfanz9jp_rg_lead_notes: Data: 0.02MB + Index: 0.03MB
wp_37gfanz9jp_signups: Data: 0.02MB + Index: 0.06MB
wp_37gfanz9jp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_37gfanz9jp_terms: Data: 0.08MB + Index: 0.09MB
wp_37gfanz9jp_term_relationships: Data: 0.16MB + Index: 0.13MB
wp_37gfanz9jp_term_taxonomy: Data: 0.08MB + Index: 0.09MB
wp_37gfanz9jp_usermeta: Data: 29.55MB + Index: 24.92MB
wp_37gfanz9jp_users: Data: 1.52MB + Index: 0.73MB
wp_37gfanz9jp_wfBadLeechers: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfBlockedCommentLog: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfBlockedIPLog: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfBlocks: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_wfBlocksAdv: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfConfig: Data: 0.33MB + Index: 0.00MB
wp_37gfanz9jp_wfCrawlers: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfFileMods: Data: 3.52MB + Index: 0.00MB
wp_37gfanz9jp_wfHits: Data: 1.02MB + Index: 0.25MB
wp_37gfanz9jp_wfHoover: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_wfIssues: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfKnownFileList: Data: 1.52MB + Index: 0.00MB
wp_37gfanz9jp_wfLeechers: Data: 2.02MB + Index: 0.00MB
wp_37gfanz9jp_wfLockedOut: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfLocs: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfLogins: Data: 0.11MB + Index: 0.03MB
wp_37gfanz9jp_wfNet404s: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_wfNotifications: Data: 0.05MB + Index: 0.00MB
wp_37gfanz9jp_wfPendingIssues: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfReverseCache: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_wfScanners: Data: 0.09MB + Index: 0.00MB
wp_37gfanz9jp_wfSNIPCache: Data: 0.02MB + Index: 0.05MB
wp_37gfanz9jp_wfStatus: Data: 0.13MB + Index: 0.11MB
wp_37gfanz9jp_wfThrottleLog: Data: 0.02MB + Index: 0.02MB
wp_37gfanz9jp_wfVulnScanners: Data: 0.02MB + Index: 0.00MB
wp_37gfanz9jp_yoast_seo_links: Data: 0.52MB + Index: 0.19MB
wp_37gfanz9jp_yoast_seo_meta: Data: 0.13MB + Index: 0.00MB

### Post Type Counts ###

attachment: 6060
leadpages_post: 9
mc4wp-form: 2
nav_menu_item: 46
page: 77
payment_retry: 1260
polylang_mo: 2
portfolio: 8
post: 282
product: 215
product_variation: 635
quiz: 2
revision: 474
scheduled-action: 2916
shop_coupon: 456
shop_order: 21552
shop_order_refund: 420
shop_subscription: 3913
shop_webhook: 30
staticblocks: 2
vc_grid_item: 3
wpsl_stores: 10

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (34) ###

Query Monitor: by John Blackbourn – 2.17.0
Gravity Forms: by rocketgenius – 2.2.5
Redux Framework: by Team Redux – 3.6.7.7
WPML Multilingual CMS: by OnTheGoSystems – 3.8.4
Akismet Anti-Spam: by Automattic – 4.0.1
CMB2: by CMB2 team – 2.2.6
Custom Sidebars: by WPMU DEV – 3.1.1
XStore Core: by 8theme – 1.0.19
Gravity Forms User Registration Add-On: by rocketgenius – 3.8
WPBakery Page Builder: by Michael M - WPBakery.com – 5.4.3
MailChimp for WordPress: by ibericode – 4.1.11
MC4WP - WPML: by ibericode – 1.0
Prevent webhook auto-disable: by Aelxandre Bouchard –  – Not tested with the active version of WooCommerce
Product Catalog Feed Pro by PixelYourSite: by PixelYourSite – 3.3.3.1
Redis Object Cache: by Till Krüss – 1.3.5
Registration Honeypot: by Justin Tadlock – 1.0.0
UpdraftPlus - Backup/Restore: by UpdraftPlus.com
DavidAnderson – 1.13.12

WooCommerce Customer/Order CSV Export: by SkyVerge – 4.4.0
WooCommerce Stripe Gateway: by WooCommerce – 3.2.3 – Not tested with the active version of WooCommerce
WooCommerce Google Analytics Integration: by WooCommerce – 1.4.3 – Not tested with the active version of WooCommerce
WooCommerce Multilingual: by OnTheGoSystems – 4.2.5
WooCommerce Smart Coupons: by WooCommerce – 3.3.8 – Not tested with the active version of WooCommerce
WooCommerce Subscriptions - Disable Report Cache Updates: by Prospress Inc. – 1.0 – Not tested with the active version of WooCommerce
WooCommerce Subscriptions: by Prospress Inc. – 2.2.15
WooCommerce: by Automattic – 3.2.4
Wordfence Security: by Wordfence – 6.3.21
Yoast SEO: by Team Yoast – 5.8
WP Facebook Pixel: by Night Shift Apps – 4.5
WP Rocket: by WP Media – 2.10.11.1
WP Store Locator: by Tijmen Smit – 2.2.9
wpMandrill: by Mandrill – 1.33
WPML Media: by OnTheGoSystems – 2.2.1
WPML String Translation: by OnTheGoSystems – 2.6.3
WPML Translation Management: by OnTheGoSystems – 2.4.3

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: CAD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
outofstock (outofstock-2)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)
rated-5 (rated-5-2)

### WC Pages ###

Shop base: #4 - /boutique/
Cart: #5 - /panier/
Checkout: #6 - /paiement/
My account: #4302 - /mon-compte/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Le Coffret de rachel
Version: 1
Author URL: https://www.kokomoweb.com
Child Theme: ✔
Parent Theme Name: XStore
Parent Theme Version: 4.13
Parent Theme Author URL: http://www.8theme.com
WooCommerce Support: ✔

### Templates ###

Overrides: xstore/woocommerce/archive-product.php
xstore/woocommerce/cart/cart-empty.php
xstore/woocommerce/cart/cart-item-data.php
xstore/woocommerce/cart/cart.php
xstore/woocommerce/cart/cross-sells.php
xstore/woocommerce/cart/mini-cart.php
xstore/woocommerce/checkout/form-billing.php
xstore/woocommerce/checkout/form-checkout.php
xstore/woocommerce/checkout/form-shipping.php
coffret-rachel/woocommerce/content-product.php
xstore/woocommerce/content-product_cat.php
coffret-rachel/woocommerce/content-single-product.php version 1.6.4 is out of date. The core version is 3.0.0
xstore/woocommerce/content-widget-product.php
coffret-rachel/woocommerce/emails/customer-completed-order.php
coffret-rachel/woocommerce/emails/customer-new-account.php
coffret-rachel/woocommerce/emails/customer-processing-order.php
coffret-rachel/woocommerce/emails/email-styles.php
xstore/woocommerce/global/breadcrumb.php
xstore/woocommerce/global/form-login.php
xstore/woocommerce/global/quantity-input.php
xstore/woocommerce/global/wrapper-end.php
xstore/woocommerce/global/wrapper-start.php
xstore/woocommerce/loop/loop-end.php
xstore/woocommerce/loop/loop-start.php
xstore/woocommerce/loop/no-products-found.php
xstore/woocommerce/loop/pagination.php
coffret-rachel/woocommerce/loop/sale-flash.php
xstore/woocommerce/single-product/add-to-cart/simple.php
coffret-rachel/woocommerce/single-product/meta.php version 1.6.4 is out of date. The core version is 3.0.0
xstore/woocommerce/single-product/product-image.php
xstore/woocommerce/single-product/related.php
xstore/woocommerce/single-product/sale-flash.php
xstore/woocommerce/single-product/short-description.php
xstore/woocommerce/single-product/tabs/additional-information.php
xstore/woocommerce/single-product/tabs/description.php
xstore/woocommerce/single-product/tabs/tabs.php
xstore/woocommerce/single-product/title.php
xstore/woocommerce/single-product/up-sells.php
xstore/woocommerce/single-product.php

Outdated Templates: ❌Learn how to update

### subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Customer/Order CSV Export ###

Order Export Format: custom
Customer Export Format: import
Automatically Export Orders: disabled
Automatically Export Customers: disabled
</details>

**Comment by mikejolley** [comment]
Related to https://github.com/woocommerce/woocommerce/pull/17748

This was brought about by the 4.9 update.

**Comment by alexbouchardd** [comment]
Missed that PR @mikejolley. Having done exactly the same changes local to test that hypothesis that indeeds seems to be the solution.

Any ideas when that will be released?

**Comment by claudiosanches** [comment]
Closed by #17748
